### PR TITLE
BigQuery trim schema with selected fields

### PR DESCRIPTION
**Pull request by RustedBones**
Extracted from https://github.com/apache/beam/pull/32360

Trim BQ schema using selected fields instead of avro schema.

### CODE REVIEW

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 1]
```diff
@@ -1441,24 +1445,6 @@ void cleanup(PassThroughThenCleanup.ContextContainer c) throws Exception {
...
-    private static Schema getFinalSchema(
-        Schema beamSchema, ValueProvider<List<String>> selectedFields) {
-      List<Schema.Field> flds =
-          beamSchema.getFields().stream()
-              .filter(
-                  field -> {
-                    if (selectedFields != null
-                        && selectedFields.isAccessible()
-                        && selectedFields.get() != null) {
-                      return selectedFields.get().contains(field.getName());
```
Comment (RustedBones) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java:1453]: selected nested field were always filtered out the beam schema

**Review by ahmedabu98 — COMMENTED** [review]
Thanks, left some comments. Generally looks good, but would like to know the motivation for this change

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQuerySourceDef.java [hunk 2]
```diff
@@ -45,12 +45,23 @@ <T> BigQuerySourceBase<T> toSource(
...
    * Extract the Beam {@link Schema} corresponding to this source.
    *
    * @param bqOptions BigQueryOptions
    * @return Beam schema of the source
    * @throws BigQuerySchemaRetrievalException if schema retrieval fails
    */
-  Schema getBeamSchema(BigQueryOptions bqOptions);
+  default Schema getBeamSchema(BigQueryOptions bqOptions) {
+    return BigQueryUtils.fromTableSchema(getTableSchema(bqOptions));
+  }
```
[the remark this answers is not in the document]
Comment (RustedBones) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQuerySourceDef.java:66]: I removed it in the original PR but was asked to put it back https://github.com/apache/beam/pull/32360#discussion_r1761144791

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryStorageSourceBase.java [hunk 3]
```diff
@@ -182,30 +178,18 @@ public List<BigQueryStorageStreamSource<T>> split(
...
-    } else if (readSession.getDataFormat() == DataFormat.AVRO) {
-      sessionSchema = new Schema.Parser().parse(readSession.getAvroSchema().getSchema());
-    } else {
-      throw new IllegalArgumentException(
-          "data is not in a supported dataFormat: " + readSession.getDataFormat());
+    // TODO: this is inconsistent with method above, where it can be null
+    Preconditions.checkStateNotNull(targetTable);
+    TableSchema tableSchema = targetTable.getSchema();
+    if (selectedFieldsProvider != null && selectedFieldsProvider.isAccessible()) {
+      tableSchema = BigQueryUtils.trimSchema(tableSchema, selectedFieldsProvider.get());
```
[the remark this answers is not in the document]
Comment (RustedBones) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryStorageSourceBase.java:185]: Yes, the operation here is correct. It is just for consistency/simplicity:

- We also trim the schema in the `BigQueryIO` where we do not have access to a trimmed avro schema.
- when we direct read with `ARROW`, to get access to the avro schema we do `arrow -> beam -> avro` schema conversions just for the sake of filtering known fields.

[the remark this answers is not in the document]
Comment (ahmedabu98) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryStorageSourceBase.java:185]: I see, thanks for explaining. Just wanted to make sure I wasn't missing something bigger.

**Review by ahmedabu98 — APPROVED** [review]
LGTM

**Pull request merged**